src/gui/qgstemporalcontrollerwidget.cpp

```diff
@@ -498,6 +508,34 @@ void QgsTemporalControllerWidget::mRangeSetToAllLayersAction_triggered()
   saveRangeToProject();
 }
 
+void QgsTemporalControllerWidget::setTimeStep( const QgsInterval &timeStep )
+{
+  if ( ! timeStep.isValid() || timeStep.seconds() <= 0 )
+    return;
+  int indexUnit = -1;
+  double value = std::numeric_limits<double>::min();
+  double previousValue;
+  do
+  {
+    indexUnit++;
+    previousValue = value;
+    QgsUnitTypes::TemporalUnit unit = static_cast<QgsUnitTypes::TemporalUnit>( mTimeStepsComboBox->itemData( indexUnit ).toInt() );
+    value = timeStep.seconds() * QgsUnitTypes::fromUnitToUnitFactor( QgsUnitTypes::TemporalSeconds, unit );
+  }
+  while ( value > 10 ||  indexUnit == mTimeStepsComboBox->count() );
+
+  if ( fabs( log( value ) ) > fabs( log( previousValue ) ) )
+  {
+    indexUnit--;
+    value = previousValue;
+  }
```

Comment:
Can you move this logic out and sit it in QgsInterval?

Reply:
The whole logic? Not sure because that depends on the units present in the combo box (not sure all unit are present).
Maybe with providing the list of unit.

Reply:
Ah - I missed that. Could we at least get some comments in here explaining the logic?

Reply:
Finally, the approach to select the time step unit has changed. Indeed, first approach was wrong and only around numerical value. Best to search with value AND significant digits.